File: go/test/projects_classic_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestProjectsClassicEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ProjectsClassic(nil)
		if ent == nil {
			t.Fatal("expected non-nil ProjectsClassicEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := projects_classicBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "projects_classic." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_PROJECTS_CLASSIC_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		projectsClassicRef01Ent := client.ProjectsClassic(nil)
		projectsClassicRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "projects_classic"}), "projects_classic_ref01"))
		projectsClassicRef01Data["column_id"] = setup.idmap["column01"]
		projectsClassicRef01Data["project_id"] = setup.idmap["project01"]

		projectsClassicRef01DataResult, err := projectsClassicRef01Ent.Create(projectsClassicRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		projectsClassicRef01Data = core.ToMapAny(entityData(projectsClassicRef01DataResult))
		if projectsClassicRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// UPDATE
		projectsClassicRef01DataUp0Up := map[string]any{
			"project_id": setup.idmap["project_id"],
		}

		projectsClassicRef01MarkdefUp0Name := "permission"
		projectsClassicRef01MarkdefUp0Value := fmt.Sprintf("Mark01-projects_classic_ref01_%d", setup.now)
		projectsClassicRef01DataUp0Up[projectsClassicRef01MarkdefUp0Name] = projectsClassicRef01MarkdefUp0Value

		projectsClassicRef01ResdataUp0Result, err := projectsClassicRef01Ent.Update(projectsClassicRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		projectsClassicRef01ResdataUp0 := core.ToMapAny(entityData(projectsClassicRef01ResdataUp0Result))
		if projectsClassicRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if projectsClassicRef01ResdataUp0[projectsClassicRef01MarkdefUp0Name] != projectsClassicRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", projectsClassicRef01MarkdefUp0Name, projectsClassicRef01ResdataUp0[projectsClassicRef01MarkdefUp0Name])
		}


	})
}

func projects_classicBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "projects_classic", "ProjectsClassicTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read projects_classic test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse projects_classic test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"projects_classic01", "projects_classic02", "projects_classic03", "column01", "column02", "column03", "project01", "project02", "project03", "collaborator01", "collaborator02", "collaborator03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_PROJECTS_CLASSIC_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_PROJECTS_CLASSIC_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
		"GITHUB_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_PROJECTS_CLASSIC_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add project_id alias for update test.
	if idmapResolved["project_id"] == nil {
		idmapResolved["project_id"] = idmapResolved["project01"]
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["GITHUB_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/projects_classic_entity_test.lua
-- ProjectsClassic entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("github_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("ProjectsClassicEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:ProjectsClassic(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = projects_classic_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "update", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "projects_classic." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_PROJECTS_CLASSIC_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local projects_classic_ref01_ent = client:ProjectsClassic(nil)
    local projects_classic_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.projects_classic"), "projects_classic_ref01"))
    projects_classic_ref01_data["column_id"] = setup.idmap["column01"]
    projects_classic_ref01_data["project_id"] = setup.idmap["project01"]

    local projects_classic_ref01_data_result, err = projects_classic_ref01_ent:create(projects_classic_ref01_data, nil)
    assert.is_nil(err)
    projects_classic_ref01_data = helpers.to_map(type(projects_classic_ref01_data_result) == 'table' and projects_classic_ref01_data_result.data_get and projects_classic_ref01_data_result:data_get() or projects_classic_ref01_data_result)
    assert.is_not_nil(projects_classic_ref01_data)

    -- UPDATE
    local projects_classic_ref01_data_up0_up = {
      ["project_id"] = setup.idmap["project_id"],
    }

    local projects_classic_ref01_markdef_up0_name = "permission"
    local projects_classic_ref01_markdef_up0_value = "Mark01-projects_classic_ref01_" .. tostring(setup.now)
    projects_classic_ref01_data_up0_up[projects_classic_ref01_markdef_up0_name] = projects_classic_ref01_markdef_up0_value

    local projects_classic_ref01_resdata_up0_result, err = projects_classic_ref01_ent:update(projects_classic_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local projects_classic_ref01_resdata_up0 = helpers.to_map(type(projects_classic_ref01_resdata_up0_result) == 'table' and projects_classic_ref01_resdata_up0_result.data_get and projects_classic_ref01_resdata_up0_result:data_get() or projects_classic_ref01_resdata_up0_result)
    assert.is_not_nil(projects_classic_ref01_resdata_up0)
    assert.are.equal(projects_classic_ref01_resdata_up0[projects_classic_ref01_markdef_up0_name], projects_classic_ref01_markdef_up0_value)


  end)
end)

function projects_classic_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/projects_classic/ProjectsClassicTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read projects_classic test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "projects_classic01", "projects_classic02", "projects_classic03", "column01", "column02", "column03", "project01", "project02", "project03", "collaborator01", "collaborator02", "collaborator03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("GITHUB_TEST_PROJECTS_CLASSIC_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["GITHUB_TEST_PROJECTS_CLASSIC_ENTID"] = idmap,
    ["GITHUB_TEST_LIVE"] = "FALSE",
    ["GITHUB_TEST_EXPLAIN"] = "FALSE",
    ["GITHUB_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["GITHUB_TEST_PROJECTS_CLASSIC_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["project_id"] == nil then
    idmap_resolved["project_id"] = idmap_resolved["project01"]
  end

  if env["GITHUB_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["GITHUB_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["GITHUB_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["GITHUB_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
